File: templates/module-auth/skeleton/src/auth/providers/apikey.ts
// ── API Key Provider ─────────────────────────────────────────────────
//
// Validates requests by checking an API key sent in the X-API-Key
// header (or Authorization: Bearer) against a set of valid keys
// configured via environment variables.
//
// Environment variables:
//   AUTH_API_KEYS — Comma-separated list of valid API keys
//
// Each key can optionally include metadata in the format:
//   key:role1+role2:label
// For example:
//   sk-abc123:admin+editor:production-key,sk-xyz789:reader:staging-key

import type { AuthResult } from "../types.js";
import type { AuthProvider, AuthRequest } from "./types.js";
import { registerProvider } from "./registry.js";

interface ParsedKey {
  key: string;
  roles: string[];
  label: string;
}

/**
 * Parse the AUTH_API_KEYS environment variable into structured key entries.
 */
function parseApiKeys(raw: string): ParsedKey[] {
  return raw
    .split(",")
    .map((entry) => entry.trim())
    .filter(Boolean)
    .map((entry) => {
      const parts = entry.split(":");
      return {
        key: parts[0]!,
        roles: parts[1]?.split("+").filter(Boolean) ?? [],
        label: parts[2] ?? "unnamed",
      };
    });
}

/**
 * Extract the API key from either the X-API-Key header or the
 * Authorization: Bearer header.
 */
function extractApiKey(request: AuthRequest): string | null {
  const getHeader = (name: string): string | null | undefined =>
    typeof request.headers.get === "function"
      ? request.headers.get(name)
      : (request.headers[name] as string | undefined);

  // Prefer X-API-Key header
  const apiKeyHeader = getHeader("x-api-key");
  if (apiKeyHeader && typeof apiKeyHeader === "string") {
    return apiKeyHeader;
  }

  // Fall back to Bearer token
  const authHeader = getHeader("authorization");
  if (authHeader && typeof authHeader === "string") {
    const match = authHeader.match(/^Bearer\s+(\S+)$/i);
    return match?.[1] ?? null;
  }

  return null;
}

const apikeyProvider: AuthProvider = {
  name: "apikey",

  async verifyRequest(request: AuthRequest): Promise<AuthResult> {
    const rawKeys = process.env.AUTH_API_KEYS;
    if (!rawKeys) {
      return {
        authenticated: false,
        error: "API key provider not configured: set AUTH_API_KEYS",
      };
    }

    const providedKey = extractApiKey(request);
    if (!providedKey) {
      return {
        authenticated: false,
        error: "Missing API key (send via X-API-Key header or Bearer token)",
      };
    }

    const validKeys = parseApiKeys(rawKeys);

    // Constant-time-ish comparison to resist timing attacks.
    // For production use, consider a proper constant-time comparison.
    const matched = validKeys.find((entry) => {
      if (entry.key.length !== providedKey.length) return false;
      let result = 0;
      for (let i = 0; i < entry.key.length; i++) {
        result |= entry.key.charCodeAt(i) ^ providedKey.charCodeAt(i);
      }
      return result === 0;
    });

    if (!matched) {
      return { authenticated: false, error: "Invalid API key" };
    }

    return {
      authenticated: true,
      user: {
        id: `apikey:${matched.label}`,
        roles: matched.roles,
        metadata: { keyLabel: matched.label, authMethod: "apikey" },
      },
    };
  },
};

// Self-register
registerProvider(apikeyProvider);

export { apikeyProvider };

File: templates/module-auth-go/skeleton/auth/providers/apikey.go
package providers

import (
	"crypto/subtle"
	"fmt"
	"net/http"
	"os"
	"strings"

	"__GO_MODULE__/auth"
)

// APIKeyProvider validates requests carrying an API key in either the
// `X-API-Key` header or an Authorization header of the form
// "ApiKey <value>". Keys are loaded from the AUTH_API_KEYS environment
// variable as comma-separated entries of the form:
//
//	key[:role1+role2][:label]
//
// Example:
//
//	AUTH_API_KEYS=sk_live_abc:admin:billing-bot,sk_live_def:reader:analytics
type APIKeyProvider struct {
	keys map[string]apiKeyEntry
}

type apiKeyEntry struct {
	roles []string
	label string
}

// NewAPIKeyProvider constructs an APIKeyProvider from AUTH_API_KEYS.
func NewAPIKeyProvider() (*APIKeyProvider, error) {
	raw := os.Getenv("AUTH_API_KEYS")
	if raw == "" {
		return nil, fmt.Errorf("%w: set AUTH_API_KEYS", ErrProviderMisconfigured)
	}
	keys := map[string]apiKeyEntry{}
	for _, part := range strings.Split(raw, ",") {
		part = strings.TrimSpace(part)
		if part == "" {
			continue
		}
		fields := strings.Split(part, ":")
		key := fields[0]
		if key == "" {
			continue
		}
		entry := apiKeyEntry{}
		if len(fields) >= 2 && fields[1] != "" {
			entry.roles = strings.Split(fields[1], "+")
		}
		if len(fields) >= 3 {
			entry.label = fields[2]
		}
		keys[key] = entry
	}
	if len(keys) == 0 {
		return nil, fmt.Errorf("%w: AUTH_API_KEYS contained no valid entries", ErrProviderMisconfigured)
	}
	return &APIKeyProvider{keys: keys}, nil
}

func (p *APIKeyProvider) Name() string { return "apikey" }

func (p *APIKeyProvider) VerifyRequest(r *http.Request) auth.Result {
	presented := extractAPIKey(r)
	if presented == "" {
		return auth.Result{Reason: "missing api key"}
	}
	// Constant-time lookup to avoid leaking key prefixes via timing.
	for key, entry := range p.keys {
		if subtle.ConstantTimeCompare([]byte(key), []byte(presented)) == 1 {
			id := entry.label
			if id == "" {
				id = keyFingerprint(key)
			}
			return auth.Result{
				Authenticated: true,
				User: &auth.User{
					ID:       id,
					Roles:    entry.roles,
					Metadata: map[string]any{"key_label": entry.label},
				},
			}
		}
	}
	return auth.Result{Reason: "invalid api key"}
}

func extractAPIKey(r *http.Request) string {
	if k := r.Header.Get("X-API-Key"); k != "" {
		return strings.TrimSpace(k)
	}
	if h := r.Header.Get("Authorization"); h != "" {
		const prefix = "ApiKey "
		if len(h) > len(prefix) && strings.EqualFold(h[:len(prefix)], prefix) {
			return strings.TrimSpace(h[len(prefix):])
		}
	}
	return ""
}

// keyFingerprint returns a short, non-reversible identifier for log
// lines so operators can correlate requests to a key without printing
// it verbatim.
func keyFingerprint(k string) string {
	if len(k) <= 6 {
		return "apikey-" + k
	}
	return "apikey-" + k[:3] + "…" + k[len(k)-3:]
}
